### ISSUE TITLE
Adding GL-AX1800 Support [WIP]

### ISSUE THREAD

**Issue by solidus1983** [solidus1983/openwrt#1]
Place to discuss about adding GL-AX1800 aka the GL-iNet Flint support to OpenWRT.

@kirdesde New Discussion as suggested

**Comment by kirdesde** [comment]
Does the initramfs image boot?

I've pushed another commit to reduce the CPU frequency as the ax1800 has a ipq6000 soc that is specified as 1.2 ghz max clock.
I've just rm the 1.4, 1.6 and 1.8 ghz opp points. But be aware, the max clock is now still slightly higher as specified (1.32 ghz)


https://github.com/kirdesde/ax1800/commit/d14519ab354a637a724bc1d03f13086753f6bf9e

**Comment by solidus1983** [comment]
Thank you for the work you are doing,  just got up but the build is failing to compile seems mkimage isn't liking a few switches will upload another error.txt file shortly.

**Comment by kirdesde** [comment]
the dtb is missing

```
FATAL ERROR: Couldn't open "/home/openwrt/openwrt/build_dir/target-aarch64_cortex-a53_musl/linux-ipq60xx_generic/linux-5.15.69/arch/arm64/boot/dts/qcom/ipq6018-gl-ax1800.dtb
```
please do a 
```
make target/linux V=s
```
to build the kernel, this should also compile the dtb or should output an error message.

**Comment by solidus1983** [comment]
`make target/linux V=s` it what produced that error.

I have done an `rm -rf openwrt` and recloned the repo and then `git checkout eap610-outdoor` so setting this up now.

**Comment by solidus1983** [comment]
[errors.txt](https://github.com/solidus1983/openwrt/files/9692307/errors.txt)

After a clean repo run, so going to uncomment the commented lines now.

**Comment by kirdesde** [comment]
Now you need a complete new buildenv (including tools, toolchain etc.)

make -j$(nproc)

If you need to clean your build again, just do a "make clean"

**Comment by solidus1983** [comment]
Thanks @kirdesde i normally run in order

`./scripts/feeds update -a
./scripts/feeds install -a
make defconfig
make nconfig
make download tools toolchain
`

Thanks for your help though. I have it now running a full make.

**Comment by solidus1983** [comment]
Right ran make

`FATAL ERROR: Couldn't open "/home/openwrt/openwrt/build_dir/target-aarch64_cortex-a53_musl/linux-ipq60xx_generic/linux-5.15.69/arch/arm64/boot/dts/qcom/ipq6018-gl-ax1800.dtb": No such file or directory
mkimage: Can't open /home/openwrt/openwrt/build_dir/target-aarch64_cortex-a53_musl/linux-ipq60xx_generic/tmp/openwrt-ipq60xx-generic-gl-ax1800-initramfs-fit-uImage.itb.new.tmp: No such file or directory
Error: Bad parameters for FIT image type
Usage: mkimage [-T type] -l image
          -l ==> list image header information
          -T ==> parse image file as 'type'
          -q ==> quiet
       mkimage [-x] -A arch -O os -T type -C comp -M magic -a addr -e ep -n name -d data_file[:data_file...] image
          -A ==> set architecture to 'arch'
          -O ==> set operating system to 'os'
          -T ==> set image type to 'type'
          -C ==> set compression type 'comp'
          -M ==> set image magic to 'magic'
          -a ==> set load address to 'addr' (hex)
          -e ==> set entry point to 'ep' (hex)
          -n ==> set image name to 'name'
          -R ==> set second image name to 'name'
          -d ==> use image data from 'datafile'
          -x ==> set XIP (execute in place)
          -s ==> create an image with no data
          -v ==> verbose
       mkimage [-D dtc_options] [-f fit-image.its|-f auto|-F] [-b <dtb> [-b <dtb>]] [-E] [-B size] [-i <ramdisk.cpio.gz>] fit-image
           <dtb> file is used with -f auto, it may occur multiple times.
          -D => set all options for device tree compiler
          -f => input filename for FIT source
          -i => input filename for ramdisk file
          -E => place data outside of the FIT structure
          -B => align size in hex for FIT structure and header
          -b => append the device tree binary to the FIT
          -t => update the timestamp in the FIT
Signing / verified boot options: [-k keydir] [-K dtb] [ -c <comment>] [-p addr] [-r] [-N engine]
          -k => set directory containing private keys
          -K => write public keys to this .dtb file
          -G => use this signing key (in lieu of -k)
          -c => add comment in signature node
          -F => re-sign existing FIT image
          -p => place external data at a static position
          -r => mark keys used as 'required' in dtb
          -N => openssl engine to use for signing
          -o => algorithm to use for signing
       mkimage -V ==> print version information and exit
Use '-T list' to see a list of available image types
make[5]: *** [Makefile:19: /home/openwrt/openwrt/build_dir/target-aarch64_cortex-a53_musl/linux-ipq60xx_generic/tmp/openwrt-ipq60xx-generic-gl-ax1800-initramfs-fit-uImage.itb] Error 1
make[4]: *** [Makefile:20: install] Error 2
make[3]: *** [Makefile:11: install] Error 2
make[2]: *** [target/Makefile:30: target/linux/install] Error 1
make[1]: *** [target/Makefile:24: /home/openwrt/openwrt/staging_dir/target-aarch64_cortex-a53_musl/stamp/.target_install] Error 2
make: *** [/home/openwrt/openwrt/include/toplevel.mk:231: world] Error 2
`
Running `make target/linux V=s` states everything is upto date.

**Comment by kirdesde** [comment]
please do a 

```
make target/linux/{clean,compile} V=s
```
and post the output.

**Comment by kirdesde** [comment]
and please post your .config

**Comment by solidus1983** [comment]
[config.txt](https://github.com/solidus1983/openwrt/files/9692694/config.txt)

**Comment by solidus1983** [comment]
[output.txt](https://github.com/solidus1983/openwrt/files/9692720/output.txt)

There is the output from the command you asked me to run.

**Comment by kirdesde** [comment]
the gl-ax-1800.dtb is missing here:

```
 DTC     arch/arm64/boot/dts/qcom/apq8016-sbc.dtb
  DTC     arch/arm64/boot/dts/qcom/apq8094-sony-xperia-kitakami-karin_windy.dtb
  DTC     arch/arm64/boot/dts/qcom/apq8096-db820c.dtb
  DTC     arch/arm64/boot/dts/qcom/apq8096-ifc6640.dtb
  DTC     arch/arm64/boot/dts/qcom/ipq6018-cp01-c1.dtb
  DTC     arch/arm64/boot/dts/qcom/ipq6018-eap610-outdoor.dtb
arch/arm64/boot/dts/qcom/ipq6018-ess.dtsi:6.3-32: Warning (reg_format): /soc/ess-switch@3a000000:reg: property has invalid length (8 bytes) (#address-cells == 2, #size-cells == 2)
arch/arm64/boot/dts/qcom/ipq6018-ess.dtsi:116.3-29: Warning (reg_format): /soc/ess-uniphy@7a00000:reg: property has invalid length (8 bytes) (#address-cells == 2, #size-cells == 2)
arch/arm64/boot/dts/qcom/ipq6018-eap610-outdoor.dtb: Warning (pci_device_reg): Failed prerequisite 'reg_format'
arch/arm64/boot/dts/qcom/ipq6018-eap610-outdoor.dtb: Warning (pci_device_bus_num): Failed prerequisite 'reg_format'
arch/arm64/boot/dts/qcom/ipq6018-eap610-outdoor.dtb: Warning (i2c_bus_reg): Failed prerequisite 'reg_format'
arch/arm64/boot/dts/qcom/ipq6018-eap610-outdoor.dtb: Warning (spi_bus_reg): Failed prerequisite 'reg_format'
  DTC     arch/arm64/boot/dts/qcom/ipq6018-mr7350.dtb
```
Please post the output of 

```
cat build_dir/target-aarch64_cortex-a53_musl/linux-ipq60xx_generic/linux-5.15.69/arch/arm64/boot/dts/qcom/Makefile
```

**Comment by solidus1983** [comment]
`cat: build_dir/target-aarch64_cortex-a53_musl/linux-ipq60xx_generic/linux-5.15.69/arch/arm64/boot/dts/qcom/Makefile: No such file or directory`

**Comment by kirdesde** [comment]
okay, the gl-ax1800 is missing here as well.

Output of:
```
cat target/linux/ipq60xx/patches-5.15/1000-arm64-dts-add-OpenWrt-DTS-files.patch
```

**Comment by solidus1983** [comment]
From 0b8ea8ba796dfeb9e9012ae07633e9df15c4c7da Mon Sep 17 00:00:00 2001
From: Robert Marko <robimarko@gmail.com>
Date: Tue, 11 May 2021 13:29:33 +0200
Subject: [PATCH 1000/1011] arm64: dts: add OpenWrt DTS files

Lets add custom OpenWrt DTS files to Makefile.

Signed-off-by: Robert Marko <robimarko@gmail.com>
[mrnuke: Replace ipq807x dtbs with ipx601x]
Signed-off-by: Alexandru Gagniuc <mr.nuke.me@gmail.com>
---
 arch/arm64/boot/dts/qcom/Makefile | 1 +
 1 file changed, 1 insertion(+)

diff --git a/arch/arm64/boot/dts/qcom/Makefile b/arch/arm64/boot/dts/qcom/Makefile
index 70516508be56..27a7e3f98465 100644
--- a/arch/arm64/boot/dts/qcom/Makefile
+++ b/arch/arm64/boot/dts/qcom/Makefile
@@ -4,6 +4,8 @@ dtb-$(CONFIG_ARCH_QCOM) += apq8094-sony-xperia-kitakami-karin_windy.dtb
 dtb-$(CONFIG_ARCH_QCOM)        += apq8096-db820c.dtb
 dtb-$(CONFIG_ARCH_QCOM)        += apq8096-ifc6640.dtb
 dtb-$(CONFIG_ARCH_QCOM)        += ipq6018-cp01-c1.dtb
+dtb-$(CONFIG_ARCH_QCOM)        += ipq6018-eap610-outdoor.dtb
+dtb-$(CONFIG_ARCH_QCOM)        += ipq6018-mr7350.dtb
 dtb-$(CONFIG_ARCH_QCOM)        += ipq8074-hk01.dtb
 dtb-$(CONFIG_ARCH_QCOM)        += ipq8074-hk10-c1.dtb
 dtb-$(CONFIG_ARCH_QCOM)        += ipq8074-hk10-c2.dtb
+dtb-$(CONFIG_ARCH_QCOM)    += ipq6018-gl-ax1800.dtb
--
2.37.1

**Comment by kirdesde** [comment]
Found the reason, my fault :-(

I pushed a fix

https://github.com/kirdesde/ax1800/commit/cdba7b762ed0a789340df7a26d5dcccb37cfe5e6

cherry-pick that, make target/linux/{clean,compile} && make

**Comment by solidus1983** [comment]
will do once the current compile completes.

I notice you say cherry pick how does one do that ? as i am converting the fixes to .patch files and running though git am.

**Comment by kirdesde** [comment]
convert to .patch + git am is also fine.

cherry-picking:

git remote add ax1800_kirdesde https://github.com/kirdesde/ax1800.git
git fetch ax1800_kirdesde
git cherry-pick cdba7b762ed0a789340df7a26d5dcccb37cfe5e6

**Comment by solidus1983** [comment]
thanks kirdesde will add that to my knowledge doc.

Running new compile run now.

**Comment by kirdesde** [comment]
I've also added a github workflow for an image build:

https://github.com/kirdesde/ax1800/actions

If that's done, you can download the images.

**Comment by solidus1983** [comment]
Thank you, you're a legend bud.

**Comment by solidus1983** [comment]
Right it made the .dtb file in the folder after that adjustment. Running main compile now.

**Comment by kirdesde** [comment]
let's see if it actually boot that

**Comment by solidus1983** [comment]
already plugging everything up now.

**Comment by solidus1983** [comment]
BOOOOOOOOOOTTTTTTTTTTT!!!!!!!!!!!!
[serial.txt](https://github.com/solidus1983/openwrt/files/9692818/serial.txt)

**Comment by kirdesde** [comment]
wireless is not working as the board-2.bin is missing right now, we will fix that later.

First we need to fix that qca-nss-dp issue, that is the reason for the missing ethernet.

I'll have a look.....

**Comment by solidus1983** [comment]
thanks bud.

**Comment by kirdesde** [comment]
okay, found something.

pushed 2 commits, cherry-pick them and do a kernel recompile && make

**Comment by solidus1983** [comment]
sweet.

**Comment by solidus1983** [comment]
[serial.txt](https://github.com/solidus1983/openwrt/files/9692871/serial.txt)

Seems ethernet is showing.

**Comment by kirdesde** [comment]
Great (but the serial.txt is still the faulty one)

Please upload the current one.

Do you get an IP on a connected client (dhcp)?

No need to pay me, it makes fun to get more devices working ;-)

**Comment by solidus1983** [comment]
[serial2.txt](https://github.com/solidus1983/openwrt/files/9692905/serial2.txt)

Here it is, also ip shows with `ip addr` however dhcp doesn't seem to be working.

I feel like i must give you something for at least your time.

**Comment by kirdesde** [comment]
Pushed 3 more commits, please try that. As usual compile new kernel first

**Comment by solidus1983** [comment]
git cherry-pick 101ea9d1f54b8b4031cf6d8e810f40b49b0d5e9f comes out with
`fatal: bad object 101ea9d1f54b8b4031cf6d8e810f40b49b0d5e9f`

**Comment by kirdesde** [comment]
have you git fetch before?

**Comment by solidus1983** [comment]
No i didn't i take it each time i want to cherry pick i have to use

git fetch ax1800_kirdesde
git cherry-pick <commit id>

**Comment by kirdesde** [comment]
Yeah, there's no automatic update of the remote repo. You always have to use git fetch if I've updated something.

**Comment by kirdesde** [comment]
And 3 more commits regarding wifi....

**Comment by kirdesde** [comment]
After cherry-picking the last commits do a "make defconfig"

**Comment by solidus1983** [comment]
Done and compiling. I take it if i was to git push from my build box it should only push to my repo right. As i have a VScode server for edits and then a sperate box for building, so i end up doing edits with vscode committing them then pull them on build box.

**Comment by kirdesde** [comment]
Yes your push is only going to your repo.

**Comment by solidus1983** [comment]
ok, why i did the setup the way i did i do not know. right new image built.

**Comment by kirdesde** [comment]
Is ethernet working now?

**Comment by kirdesde** [comment]
Can you check in menuconfig if ath11k-firmware-ipq6018 and ath11k-bdf-ipq6018 is selected (under firmware)?

**Comment by solidus1983** [comment]
Will do and also dhcp is not working on ethernet.

**Comment by kirdesde** [comment]
please post also the output of "ip a"

**Comment by solidus1983** [comment]
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
2: eth0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc mq master br-lan state DOWN qlen 1000
    link/ether 42:59:be:4a:7a:21 brd ff:ff:ff:ff:ff:ff
3: eth1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq master br-lan state UP qlen 1000
    link/ether 72:0a:24:14:b5:fe brd ff:ff:ff:ff:ff:ff
4: eth2: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc mq master br-lan state DOWN qlen 1000
    link/ether 72:a5:53:cc:4e:d7 brd ff:ff:ff:ff:ff:ff
5: eth3: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc mq master br-lan state DOWN qlen 1000
    link/ether 02:5e:07:c4:6a:96 brd ff:ff:ff:ff:ff:ff
6: eth4: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc mq state DOWN qlen 1000
    link/ether 3a:83:f1:3e:3f:f5 brd ff:ff:ff:ff:ff:ff
7: br-lan: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP qlen 1000
    link/ether 42:59:be:4a:7a:21 brd ff:ff:ff:ff:ff:ff
    inet 192.168.1.1/24 brd 192.168.1.255 scope global br-lan
       valid_lft forever preferred_lft forever
    inet6 fd9a:b69f:17f2::1/60 scope global noprefixroute
       valid_lft forever preferred_lft forever
    inet6 fe80::4059:beff:fe4a:7a21/64 scope link
       valid_lft forever preferred_lft forever
root@OpenWrt:/#

**Comment by kirdesde** [comment]
please also try a fixed ip on the client (192.168.1.2 for example)

**Comment by solidus1983** [comment]
Pinging 192.168.1.1 with 32 bytes of data:
Reply from 192.168.1.88: Destination host unreachable.
Reply from 192.168.1.101: Destination host unreachable.
Reply from 192.168.1.101: Destination host unreachable.
Reply from 192.168.1.101: Destination host unreachable.

Ping statistics for 192.168.1.1:
    Packets: Sent = 4, Received = 4, Lost = 0 (0% loss),

**Comment by kirdesde** [comment]
Why are there t different IP's in the reply output?

Do you have another network connected with the same subnet?

**Comment by kirdesde** [comment]
BTW I pushed another commit.

**Comment by solidus1983** [comment]
Yeah i forgot to turn off my wireless from main network.

Pinging 192.168.1.1 with 32 bytes of data:
Reply from 192.168.1.88: Destination host unreachable.
Reply from 192.168.1.88: Destination host unreachable.
Reply from 192.168.1.88: Destination host unreachable.
Reply from 192.168.1.88: Destination host unreachable.

Ping statistics for 192.168.1.1:
    Packets: Sent = 4, Received = 4, Lost = 0 (0% loss),

**Comment by kirdesde** [comment]
try the new commit and make sure the mentioned ath11k packages are select and do a new build please

**Comment by kirdesde** [comment]
And it's good practise to use another subnet on the device, just do a 
```
vi /etc/config/network
-> change the ip on the br-lan
:x
/etc/init.d/network restart
```
Due to running from initramfs you have to do this after each bootm

**Comment by solidus1983** [comment]
yeah totally forgot to do that, its a my bad thing. Compile is rolling.

**Comment by solidus1983** [comment]
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
2: eth0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc mq state DOWN qlen 1000
    link/ether d0:0d:fe:ed:01:32 brd ff:ff:ff:ff:ff:ff
3: eth1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq master br-lan state UP qlen 1000
    link/ether 2a:d4:00:00:00:38 brd ff:ff:ff:ff:ff:ff
4: eth2: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc mq master br-lan state DOWN qlen 1000
    link/ether da:e9:37:93:46:9d brd ff:ff:ff:ff:ff:ff
5: eth3: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc mq master br-lan state DOWN qlen 1000
    link/ether 76:22:67:05:46:d3 brd ff:ff:ff:ff:ff:ff
6: eth4: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc mq master br-lan state DOWN qlen 1000
    link/ether a6:f2:e4:02:10:54 brd ff:ff:ff:ff:ff:ff
7: wlan0: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN qlen 4096
    link/ether 00:03:7f:12:1b:07 brd ff:ff:ff:ff:ff:ff
8: wlan1: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN qlen 4096
    link/ether 00:03:7f:12:23:8f brd ff:ff:ff:ff:ff:ff
9: br-lan: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP qlen 1000
    link/ether 2a:d4:00:00:00:38 brd ff:ff:ff:ff:ff:ff
    inet 192.168.1.1/24 brd 192.168.1.255 scope global br-lan
       valid_lft forever preferred_lft forever
    inet6 fd9c:f52a:c973::1/60 scope global noprefixroute
       valid_lft forever preferred_lft forever
    inet6 fe80::28d4:ff:fe00:38/64 scope link
       valid_lft forever preferred_lft forever

**Comment by kirdesde** [comment]
please post also the serial.txt

**Comment by solidus1983** [comment]
[serial5.txt](https://github.com/solidus1983/openwrt/files/9693047/serial5.txt)

Wireless DHCP's fine and pings just fine.

**Comment by kirdesde** [comment]
Okay, so still no ethernet?
let me think about it

**Comment by solidus1983** [comment]
Yes no ethernet at the moment, will let you think about and take your time. Thanks for the help.

**Comment by kirdesde** [comment]
Can you eventually try if the wan port is working? Just connect to another router lan port and look if the ax1800 is getting a wan ip.

**Comment by solidus1983** [comment]
Yeah sure let me get out my GL-AR750 and its battery power pack.

**Comment by kirdesde** [comment]
I found something and pushed a new commit, please try that.

**Comment by solidus1983** [comment]
Will do that now.

**Comment by solidus1983** [comment]
Right its reboot looping
[serial6.txt](https://github.com/solidus1983/openwrt/files/9693224/serial6.txt)

Will redownload the git repo and try later going to have a few games thanks for your help.

**Comment by kirdesde** [comment]
have you power-cycled the board?

**Comment by solidus1983** [comment]
Yes

**Comment by kirdesde** [comment]
Can you post the output  of:

```
 cat /sys/kernel/debug/gpio
 ```
 
 Are the led`s working?

**Comment by solidus1983** [comment]
I don't get a chance to get to a shell before it reboots.

**Comment by kirdesde** [comment]
Yes, that's right. Can you revert the last commit, rebuild and check again and post the output?

**Comment by solidus1983** [comment]
Will do once I finish my gaming session

**Comment by kirdesde** [comment]
Please cherry-pick the latest commit and try that. Something is not right with the phy reset

**Comment by solidus1983** [comment]
Do i still need to revert the last commit before the new one ?

**Comment by solidus1983** [comment]
Right latest commit runs without the reset issue.
[serial7.txt](https://github.com/solidus1983/openwrt/files/9693619/serial7.txt)

Nothing showing on WAN port on AR-750's client list also still unable to ping either end.

**Comment by kirdesde** [comment]
your "fixing repo" commit has overwritten my network ports change (10f69929b18a60c0785435219d5a876f70a78659)

Can you try my images?
https://github.com/kirdesde/ax1800/actions/runs/3171640042
If it's still crashing from initramfs, can you flash the factory.ubi trough u-boot webinterface and test that?

**Comment by solidus1983** [comment]
Morning just got up, will boot main laptop up.

Which file was broken with my mistake? I take it the 02_network file which I accidentally deleted like a tool. Will look on your repo to see if I can fix it. Would forking from your repo be easier?

**Comment by kirdesde** [comment]
Morning, yeah forking my repo would makes it easier.

This file is wrong:

```
https://github.com/solidus1983/openwrt/blob/eap610-outdoor/target/linux/ipq60xx/base-files/etc/board.d/02_network
```
Should be:
```

. /lib/functions/uci-defaults.sh
. /lib/functions/system.sh

setup_network()
{
	local macaddr

	case $(board_name) in
	tplink,eap610-outdoor)
		# /tmp/factory_data should be mounted by preinit
		macaddr=$(get_mac_binary /tmp/factory_data/default-mac 0)
		ucidef_set_interface_macaddr "lan" "$macaddr"
		ucidef_set_interface_lan "eth0" "dhcp"
		;;
	glinet,ax1800)
		ucidef_set_interfaces_lan_wan "eth0 eth1 eth2 eth3" "eth4"
		;;
	*)
		ucidef_set_interface_lan "eth0"
		;;
	esac
}

board_config_update
setup_network
board_config_flush

exit 0
```

**Comment by solidus1983** [comment]
Will fork your repo after deleting this repo.
